dev-tools/xtask/src/illumos.rs

```diff
@@ -0,0 +1,144 @@
+    errors: &mut HashMap<String, Vec<LibraryError>>,
+) -> Result<()> {
+    let binary = path.file_name().context("basename of executable")?;
+
+    let command = Command::new("elfedit")
+        .arg("-o")
+        .arg("simple")
+        .arg("-r")
+        .arg("-e")
+        .arg("dyn:tag NEEDED")
+        .arg(&path)
+        .output()
+        .context("exec elfedit")?;
+
+    if !command.status.success() {
+        bail!("Failed to execute elfedit successfully {}", command.status);
+    }
+
+    let stdout = String::from_utf8(command.stdout)?;
+    // `elfedit -o simple -r -e "dyn:tag NEEDED" /file/path` will return
+    // a new line seperated list of required libraries so we walk over
+    // them looking for a match in our configuration file. If we find
+    // the library we make sure the binary is allowed to pull it in via
+    // the whitelist.
+    for library in stdout.lines() {
+        let library_config = match config.libraries.get(library.trim()) {
+            Some(config) => config,
+            None => {
+                errors
+                    .entry(binary.to_string())
+                    .or_default()
+                    .push(LibraryError::Unexpected(library.to_string()));
+
+                continue;
+            }
+        };
+
+        if let Some(allowed) = &library_config.binary_allow_list {
+            if !allowed.contains(binary) {
+                errors
+                    .entry(binary.to_string())
+                    .or_default()
+                    .push(LibraryError::NotAllowed(library.to_string()));
+            }
+        }
+    }
+
+    Ok(())
+}
+pub fn cmd_verify_libraries() -> Result<()> {
+    let metadata = load_workspace()?;
+    let mut config_path = metadata.workspace_root;
+    config_path.push(".cargo/xtask.toml");
+    let config = read_xtask_toml(&config_path)?;
+
+    let mut command = Command::new("cargo")
+        .args(&["build", "--message-format=json-render-diagnostics"])
+        .stdout(Stdio::piped())
+        .spawn()
+        .context("failed to spawn cargo build")?;
+
+    let reader = BufReader::new(command.stdout.take().context("take stdout")?);
+
+    let mut errors = Default::default();
+    for message in cargo_metadata::Message::parse_stream(reader) {
+        match message? {
+            Message::CompilerArtifact(artifact) => {
+                // We are only interested in artifacts that are binaries
+                if let Some(executable) = artifact.executable {
+                    verify_executable(&config, &executable, &mut errors)?;
+                }
+            }
+            _ => (),
+        }
+    }
+
+    let status = command.wait()?;
+    if !status.success() {
+        bail!("Failed to execute cargo build successfully {}", status);
+    }
+
+    if !errors.is_empty() {
+        let mut msg = String::new();
+        use std::fmt::Write;
+        errors.iter().for_each(|(binary, errors)| {
+            write!(msg, "{binary}\n").unwrap();
```

Comment:
`writeln!`

Also you can use https://crates.io/crates/swrite to avoid the unwrap.

Reply:
Fixed in [fa5b905](https://github.com/oxidecomputer/omicron/pull/5259/commits/fa5b905ef38a8a5f50130329967961514346ec67)